ios-base/Networking/Services/BaseAPIClient.swift

```diff
@@ -63,69 +54,75 @@ internal final class BaseAPIClient: APIClient {
       media: media
     ) { [weak self] result in
       guard let self = self else { return }
-
-      switch result {
-      case.success(let response):
-        self.validateResult(
-          response: response,
-          customDecodingConfiguration: endpoint.decodingConfiguration,
-          completion: completion
-        )
-      case .failure(let error):
-        completion(.failure(error), [:])
-      }
+      
+      self.handle(result, for: endpoint, completion: completion)
     }
   }
-
-  private func handleCustomAPIError(from response: Network.Response) -> APIError? {
-    if response.statusCode == Network.StatusCode.unauthorized {
-      AppDelegate.shared.unexpectedLogout()
+  
+  private func handle<T: Decodable>(
+    _ result: Result<Network.Response, Error>,
+    for endpoint: Endpoint,
+    completion: CompletionCallback<T>
+  ) {
+    switch result {
+    case .success(let response): handle(response, with: endpoint.decodingConfiguration, completion: completion)
+    case .failure(let error): completion(.failure(error), [:])
     }
-
-    return APIError(response: response, decodingConfiguration: decodingConfiguration)
   }
   
-  private func validateResult<T: Decodable>(
-    response: Network.Response,
-    customDecodingConfiguration: DecodingConfiguration?,
+  private func buildApiEndpoint(from endpoint: Endpoint) -> APIEndpoint {
+    APIEndpoint(endpoint: endpoint, headersProvider: headersProvider)
+  }
+  
+  private func handle<T: Decodable>(
+    _ response: Network.Response,
+    with configuration: DecodingConfiguration?,
     completion: CompletionCallback<T>
   ) {
-    let responseData = response.data
-
-    guard let data = responseData, !data.isEmpty else {
-      if emptyDataStatusCodes.contains(response.statusCode) {
-        completion(.success(.none), response.headers)
-      } else {
-        let emptyResponseInvalidError = App.error(
-          domain: .network,
-          localizedDescription: "Unexpected empty response".localized
-        )
-        completion(.failure(emptyResponseInvalidError), response.headers)
-      }
-
-      return
-    }
-
-    let decoder = JSONDecoder(
-      decodingConfig: customDecodingConfiguration ?? decodingConfiguration
-    )
     do {
-      let decodedObject = try decoder.decode(T.self, from: data)
+      guard let data = response.data, !data.isEmpty else {
+        try checkForUnexpectedCodeOnEmptyData(response)
+        
+        return completion(.success(.none), response.headers)
+      }
 
-      completion(.success(decodedObject), response.headers)
+      completion(.success(try decode(data, with: configuration)), response.headers)
     } catch let error {
       completion(
         .failure(handleCustomAPIError(from: response) ?? error),
         response.headers
       )
     }
   }
+  
+  private func checkForUnexpectedCodeOnEmptyData(_ response: Network.Response) throws {
+    if !emptyDataStatusCodes.contains(response.statusCode) {
```

Comment:
I think there is a semantic issue with this function, and is that it doesn't really check for the data to be empty.
IMO this function should also check about the data being empty, because the condition to show an "unexpected empty response" is both(the data being empty and the status code being in that range).

Based on the name, I would expect that but it is not the case. So I could call this function without checking for the data first and that would lead to bugs.

Reply:
I agree with what you are saying, it's kind of complex what we are trying to achieve there.

i'm going to leave as it was before.